src/connector/src/source/filesystem/opendal_source/opendal_reader.rs

```diff
@@ -215,3 +229,60 @@ impl<Src: OpendalSource> OpendalReader<Src> {
         }
     }
 }
+
+/// Extracts valid column indices from a Parquet file schema based on the user's requested schema.
+///
+/// This function is used for column pruning of Parquet files. It calculates the intersection
+/// between the columns in the currently read Parquet file and the schema provided by the user.
+/// This is useful for reading a `RecordBatch` with the appropriate `ProjectionMask`, ensuring that
+/// only the necessary columns are read.
+///
+/// # Parameters
+/// - `columns`: A vector of `Column` representing the user's requested schema.
+/// - `metadata`: A reference to `FileMetaData` containing the schema and metadata of the Parquet file.
+///
+/// # Returns
+/// - A `ConnectorResult<Vec<usize>>`, which contains the indices of the valid columns in the
+///   Parquet file schema that match the requested schema. If an error occurs during processing,
+///   it returns an appropriate error.
+pub fn extract_valid_column_indices(
+    columns: Option<Vec<Column>>,
+    metadata: &FileMetaData,
+) -> ConnectorResult<Vec<usize>> {
+    match columns {
+        Some(rw_columns) => {
+            let parquet_column_names = metadata
+                .schema_descr()
+                .columns()
+                .iter()
+                .map(|c| c.name())
+                .collect_vec();
+
+            let converted_arrow_schema =
+                parquet_to_arrow_schema(metadata.schema_descr(), metadata.key_value_metadata())
+                    .map_err(anyhow::Error::from)?;
+
+            let mut valid_column_indices: Vec<usize> = rw_columns
+                .iter()
+                .filter_map(|column| {
+                    parquet_column_names
+                        .iter()
+                        .position(|&name| name == column.name)
+                        .and_then(|pos| {
+                            let arrow_field = IcebergArrowConvert
+                                .to_arrow_field(&column.name, &column.data_type)
+                                .ok()?;
+                            if &arrow_field == converted_arrow_schema.field(pos) {
+                                Some(pos)
+                            } else {
+                                None
+                            }
+                        })
+                })
+                .collect();
+            valid_column_indices.sort();
```

Comment:
It seems the order of the input columns isn't respected, but the order matters. Considering parquet schema is ['a', 'b', 'c'], but the column order we want is [`c`, `b`]. The `valid_column_indices.sort()` here will lead to [`b`, `c`] which is unexpected.

Reply:
The `valid_column_indices` is used to construct `ProjectionMask`, and in  `ProjectionMask`, repeated or out of order indices will not impact the final mask. i.e. `[0, 1, 2]` will construct the same mask as `[1, 0, 0, 2]`. 
So no matter what order, there will be no problem with correctness. But I don't mind removing the sort here.

Reply:
Previously I keep `.sort()` just refer to how [iceberg-rust](https://github.com/apache/iceberg-rust/blob/cafc939d6aa303f7424a74fa8467ac0cb306591d/crates/iceberg/src/arrow/reader.rs#L378) does, let me also send a pr to remove this in iceberg-rust😄.

Reply:
I see. `ParquetParser` will reorder the chunk returned from `ParquetRecordBatchStream`.